### PR TITLE
ES5 distribution on npm

### PR DESCRIPTION
**Pull request by holgerd77**
Addresses #271, (as discussed and agreed upon in the PR discussion) also alternative solution to PR #274 .

This PR uses ``Babel`` to build an ``ES5`` compatible distribution build in the ``dist`` folder. The ``dist/index.js`` file is then referenced as the main entry point in ``package.js``. The ``dist`` folder itself is excluded from the repo (through ``.gitignore``) and just created before publish or testing.

The normal test runs are still done against the ``lib`` folder. There is a new test option ``--dist`` introduced, which is used for the ``package.json`` test scripts and allow for testing against the ``dist`` folder files.

~Since only the ``dist`` folder is included in the npm build, coverage is now also run against the ``dist`` folder and coveralls test report will show the ``dist`` folder files. This shouldn't be that much of a problem though, since the structure of the files didn't change a lot (except some replacements of ``const`` with ``var`` e.g.), so it should still be easily possible to identify the code parts missing code coverage.~

Update: After some experimentation it turned out that ``Coveralls`` showing the files covered couldn't be combined with the desired property of not having the ``dist`` folder in Git by excluding the folder in ``.gitignore``. So ``coverage`` is now run against the ``lib`` folder. Which also makes sense.

One note for review: I also switched the order of some config elements in ``package.json`` to be more conform with other libraries. This was a bit unlucky since it hides a bit the relevant changes when just looking at the complete code base changes. Review can therefor better be done by looking through the single commit changes.

### CODE REVIEW

**Comment by holgerd77** [comment]
I've also integrated here state tests and coverage run on travis, so that we don't have to run the state tests twice.

**Comment by holgerd77** [comment]
@Firescar96 reviewed this and gave his ok [here](https://github.com/ethereumjs/ethereumjs-vm/pull/271#issuecomment-370090402), will take this as informal review and merge.

File: .travis.yml [hunk 1]
```diff
@@ -4,7 +4,8 @@ os:
 
 language: node_js
 node_js:
-  - "6"
```
Comment (axic) [inline .travis.yml]: Why did we drop 6?

File: tests/tester.js [hunk 2]
```diff
@@ -55,7 +55,8 @@ const skipSlow = [
   'static_Call1MB1024Calldepth', // slow
   'static_Callcode50000', // slow
   'static_Return50000', // slow
-  'static_Return50000_2' // slow
+  'static_Return50000_2', // slow
+  'QuadraticComplexitySolidity_CallDataCopy'
```
Comment (axic) [inline tests/tester.js]: Why was this removed?

File: package.json [hunk 3]
```diff
@@ -1,8 +1,35 @@
...
+    "coveralls": "npm run coverage && coveralls <coverage/lcov.info",
+    "testVM": "node ./tests/tester -v",
+    "testState": "npm run build:dist && node ./tests/tester -s --dist",
+    "testBuildIntegrity": "npm run build:dist && node ./tests/tester -s --dist --test='stackOverflow'",
+    "testBlockchain": "npm run build:dist && node --stack-size=1500 ./tests/tester -b --dist --excludeDir='GeneralStateTests'",
+    "testBlockchainGeneralStateTests": "npm run build:dist && node --stack-size=1500 ./tests/tester -b --dist --dir='GeneralStateTests'",
+    "testBlockchainBlockGasLimit": "npm run build:dist && node --stack-size=1500 ./tests/tester -b --dist --dir='bcBlockGasLimitTest'",
+    "testBlockchainValid": "npm run build:dist && node --stack-size=1500 ./tests/tester -b --dist --dir='bcValidBlockTest'",
+    "testBlockchainTotalDifficulty": "npm run build:dist && node --stack-size=1500 ./tests/tester -b --dist --dir='bcTotalDifficultyTest'",
+    "test": "node ./tests/tester -a",
```
Comment (axic) [inline package.json]: This is lacking `build:dist`.

File: package.json [hunk 4]
```diff
@@ -1,8 +1,35 @@
...
   "version": "2.3.3",
-  "description": "an ethereum VM implementation",
-  "main": "index.js",
+  "description": "An Ethereum VM implementation",
+  "main": "dist/index.js",
+  "files": [
+    "dist"
+  ],
+  "scripts": {
+    "coverage": "istanbul cover ./tests/tester.js -- -s",
```
Comment (axic) [inline package.json]: This is lacking `build:dist`.

File: package.json [hunk 5]
```diff
@@ -1,8 +1,35 @@
...
+    "coveralls": "npm run coverage && coveralls <coverage/lcov.info",
+    "testVM": "node ./tests/tester -v",
```
Comment (axic) [inline package.json]: This is lacking `build:dist`.

File: package.json [hunk 6]
```diff
@@ -1,8 +1,35 @@
...
+    "testState": "npm run build:dist && node ./tests/tester -s --dist",
+    "testBuildIntegrity": "npm run build:dist && node ./tests/tester -s --dist --test='stackOverflow'",
+    "testBlockchain": "npm run build:dist && node --stack-size=1500 ./tests/tester -b --dist --excludeDir='GeneralStateTests'",
+    "testBlockchainGeneralStateTests": "npm run build:dist && node --stack-size=1500 ./tests/tester -b --dist --dir='GeneralStateTests'",
+    "testBlockchainBlockGasLimit": "npm run build:dist && node --stack-size=1500 ./tests/tester -b --dist --dir='bcBlockGasLimitTest'",
+    "testBlockchainValid": "npm run build:dist && node --stack-size=1500 ./tests/tester -b --dist --dir='bcValidBlockTest'",
+    "testBlockchainTotalDifficulty": "npm run build:dist && node --stack-size=1500 ./tests/tester -b --dist --dir='bcTotalDifficultyTest'",
+    "test": "node ./tests/tester -a",
+    "lint": "standard",
```
Comment (axic) [inline package.json]: Will `lint` ignore the `dist` directory?

File: package.json [hunk 7]
```diff
@@ -1,8 +1,35 @@
...
+    "prepublishOnly": "npm run lint && npm run build:dist && npm run testBuildIntegrity",
```
Comment (axic) [inline package.json]: Not sure if this will run `build:dist` twice?

File: .travis.yml (hunk 1 above)
Comment (holgerd77) [reply .travis.yml]: Node 6 is not dropped, it's still below in the test matrix settings. Only the ``testBuildIntegrity`` test is run with Node 8 and 9, all the other tests are run against Node 6. So all the test from before are still running with the same environment, with ``testBuildIntegrity`` (just running a single state test) we are able to do additional checks without using too much of the limited run time for tests in travis.

File: package.json (hunk 4 above)
Comment (holgerd77) [reply package.json]: No, it's not, this is explained above. I'll cite here:

"Update: After some experimentation it turned out that Coveralls showing the files covered couldn't be combined with the desired property of not having the dist folder in Git by excluding the folder in .gitignore. So coverage is now run against the lib folder. Which also makes sense."

File: package.json (hunk 6 above)
Comment (holgerd77) [reply package.json]: Yes.

File: package.json (hunk 5 above)
Comment (holgerd77) [reply package.json]: VM tests are not active atm, this would only be pseudo-consistency.

File: package.json (hunk 3 above)
Comment (holgerd77) [reply package.json]: Tests don't necessarily have to be run against the ``dist`` folder, this can now be switched providing the ``dist`` option. The ``test`` command (which I think isn't used anyhow), is remaining on `lib`` (was undecided on this). We could generally make this more explicit by renaming test commands to e.g. ``testBlockchain:dist``, not sure if this is necessary though.

File: tests/tester.js (hunk 2 above)
Comment (holgerd77) [reply tests/tester.js]: This test is extremely slow, travis is getting on its limits (sometimes).

File: package.json (hunk 7 above)
Comment (holgerd77) [reply package.json]: Yes, it does, could probably be removed. The dist build is not taking that much time though, just a couple of seconds.

**Comment by holgerd77** [comment]
@axic General remark: ok, I see, won't use external reviews as full reviews any more. Thought that would suffice (since you guys are all pretty busy atm), sorry. Next time I'll request a classic team dev review again.

File: .travis.yml (hunk 1 above)
Comment (axic) [reply .travis.yml]: Cool, travis config can be misleading in cases, especially when the section shown by github is small :)

Comment (axic) [reply .travis.yml]: So the goal of `testBuildIntegrity` is to have a quick check with a single test?

File: package.json (hunk 4 above)
Comment (axic) [reply package.json]: Ah I see, if `--dist` is specified it runs on the `dist` directory, otherwise on `lib`.

File: .travis.yml (hunk 1 above)
Comment (holgerd77) [reply .travis.yml]: Yes, to catch stuff like "is Node 10" generally working, "is the dist folder existing", stuff like that.

File: package.json (hunk 6 above)
Comment (axic) [reply package.json]: Yeah I see the `dist/**` line in a`package.json`, just the change view on github didn't made it apparent.

**Comment by axic** [comment]
In general I wanted to understand the changes, they all look ok after clarifications above.
